### PR TITLE
Fix x86 disassembly issues related to MASM syntax

### PR DESCRIPTION
**Pull request by antonangeloff**
**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [X] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [X] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Seems MASM uses "trailing-h" notation for hexadecimal integer literals as well as "0x" notation ([docs](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/masm-numbers-and-operators#numbers-in-debugger-masm-expressions)). And for some reason the disassembler outputs trailing-h, but it's not handled properly by Rizin.

The PR adds code for handling hex literals in trailing-h notation.

**Test plan**
- Added test for parsing hex literals (lowercase and uppercase)
- Added test for variable name substitution (lowercase and uppercase)
- Added test for function name substitution (lowercase  and uppercase)
- Added test for rip-related address substitution (lowercase  and uppercase)

**Closing issues**

https://github.com/rizinorg/rizin/issues/1776

### CODE REVIEW

**Comment by antonangeloff** [comment]
I've noticed an uppercase issue which is not limited to MASM syntax, here is an example with Intel:
```
[0x10000108a]> e asm.syntax=intel
[0x10000108a]> e asm.ucase=false
[0x10000108a]> pd 1
  0x10000108a      lea   rsi, section.4.__TEXT.__cstring
[0x10000108a]> e asm.ucase=true
[0x10000108a]> pd 1
  0x10000108a      LEA   RSI, [section.4.__TEXT.__cstring]
```
I.e. the square brackets are not removed. Seems that's a result of case sensitive string comparison - https://github.com/rizinorg/rizin/blob/544905631c82a52240accf5183d6f9019aaf4fb5/librz/parse/filter.c#L213

File: librz/util/str.c [hunk 1]
```diff
@@ -3122,6 +3122,14 @@ RZ_API bool rz_str_startswith(const char *str, const char *needle) {
 	return !strncmp(str, needle, strlen(needle));
 }
 
+RZ_API bool rz_str_startswith_icase(const char *str, const char *needle) {
```
Comment (ret2libc) [inline librz/util/str.c]: Add doxygen documentation please

**Review by wargio — APPROVED** [review]
LGTM

File: librz/util/str.c (hunk 1 above)
Comment (wargio) [reply librz/util/str.c]:
```suggestion
RZ_API bool rz_str_startswith_icase(RZ_NONNULL const char *str, RZ_NONNULL const char *needle) {
```

Comment (wargio) [reply librz/util/str.c]: also this change here and in the headers

Comment (antonangeloff) [reply librz/util/str.c]: Good points, added

File: librz/include/rz_util/rz_str.h [hunk 2]
```diff
@@ -236,8 +236,22 @@ RZ_API char *rz_str_replace_thunked(char *str, char *clean, int *thunk, int clen
...
 RZ_API char *rz_str_between(const char *str, const char *prefix, const char *suffix);
-RZ_API bool rz_str_startswith(const char *str, const char *needle);
-RZ_API bool rz_str_startswith_icase(const char *str, const char *needle);
+/**
+ * \brief Checks if a string starts with a specifc sequence of characters (case sensitive)
+ * \param str C-string to be scanned
+ * \param needle C-string containing the sequence of characters to match
+ * \return True if \p needle is found at the beginning of \p str and false otherwise
+ * \see rz_str_startswith_icase()
+ */
```
Comment (wargio) [inline librz/include/rz_util/rz_str.h]: these doxygen definitions goes in the C file not the header.

File: librz/include/rz_util/rz_str.h [hunk 3]
```diff
@@ -236,8 +236,22 @@ RZ_API char *rz_str_replace_thunked(char *str, char *clean, int *thunk, int clen
...
+RZ_API bool rz_str_startswith(RZ_NONNULL const char *str, RZ_NONNULL const char *needle);
+/**
+ * \brief Checks if a string starts with a specifc sequence of characters (case insensitive)
+ * \param str C-string to be scanned
+ * \param needle C-string containing the sequence of characters to match
+ * \return True if \p needle is found at the beginning of \p str and false otherwise
+ * \see rz_str_startswith()
+ */
```
Comment (wargio) [inline librz/include/rz_util/rz_str.h]: same

**Review by wargio — CHANGES_REQUESTED** [review]

**Comment by wargio** [comment]
BTW, great job :)

File: librz/include/rz_util/rz_str.h (hunk 3 above)
Comment (antonangeloff) [reply librz/include/rz_util/rz_str.h]: Okay. I've noticed some doxygen comments were placed in the headers for some functions and in the c files for other functions, I also didn't saw a recommendation regarding this in the [coding style](https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#documentation). Is there any specific rationale for placing them in the source file btw?

Comment (wargio) [reply librz/include/rz_util/rz_str.h]: we have decided to place them in the C files. only structs in the headers contains the doxygen comments.

Comment (antonangeloff) [reply librz/include/rz_util/rz_str.h]: I see, I meant I was interested to know the upsides/pros that lead the team to choose this particular style. Mostly of curiosity. Anyhow, will move them shortly.

Comment (antonangeloff) [reply librz/include/rz_util/rz_str.h]: Also @wargio @ret2libc let me know if you think it's worth adding a line to the developer guide regarding the doxygen comments location

Comment (ret2libc) [reply librz/include/rz_util/rz_str.h]: IIRC the reason was that having the doxygen doc in the headers would make them too hard to navigate with all the text. At this point I think that being consistent is more important than anything else.

Comment (antonangeloff) [reply librz/include/rz_util/rz_str.h]: I see

**Review by XVilka — APPROVED** [review]

**Review by ret2libc — APPROVED** [review]

**Pull request merged**